### PR TITLE
[FEAT] Device Credential Service

### LINKED ISSUE #64: [FEAT] DeviceCredentialService
# Contents

DeviceCredentialService를 구현한다.

## Description

user의 device별로 로그인 정보와 토큰을 저장한다. (DeviceCredential)
DeviceCredential의 도메인 로직을 활용하여, 요청에 대한 응답과 db와의 상호작용을 담당하는 응용 서비스이다.

## Todo

-   [ ] get
-   [ ] isLogin
-   [ ] login
-   [ ] logout
-   [ ] reissue

## Check Priority

Expected Priority: MEDIUM

## Reference (Optional)

(null)

### PR DESCRIPTION
**Pull request by sueelly**
## #️⃣ Related Issue

Linked Feature backlog: close #64 

## 📝 Purpose of PR

Device Credential Service를 구현한다.

## Contents

Device Credential 데이터를 다루는 Service이다.
Device Credential은 디바이스별 로그인 상태와 토큰 정보를 저장한다.
DeviceCredentialService는 DeviceCredential을 호출하여 도메인 로직을 수행하고, 수행한 결과를 저장 및 조회하는 역할이다.
- 로그인 시 LoginStatus.LOGIN으로 업데이트 및 refresh token 저장
- 로그아웃 시 LoginStatus.LOGOUT으로 업데이트 및 refresh token 삭제
- 토큰 재발급 시 토큰 업데이트

## Checklist

-   [x] Does commit messages follow the convention?
-   [x] Are variable names brief but descriptive?
-   [x] Is the code clean and easy to understand?
-   [x] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
-   [ ] Does the test successfully test edge and corner cases?

## 💬 Review Requirements (Optional)

Is there anything you can praise about this PR? Start with praise.
Make a review, leaving kind comments and suggesting changes where needed (to resolve the above).

### CODE REVIEW

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/JwtProvider.java [hunk 1]
```diff
@@ -34,21 +34,25 @@ public Token createToken(TokenPayload payload) {
...
 
     public String createAccessToken(TokenPayload payload) {
+        payload.setExpiredAt(new Date(System.currentTimeMillis() + accessTokenExpiration));
+
         return Jwts.builder()
             .setSubject(payload.getUserId().toString())
             .addClaims(Map.of("deviceId", payload.getDeviceId()))
             .setIssuedAt(payload.getIssuedAt())
-            .setExpiration(new Date(System.currentTimeMillis() + accessTokenExpiration))
+            .setExpiration(payload.getExpiredAt())
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/JwtProvider.java:43]: 여기는 왜 굳이 payload에 존재하지 않는 임의의 값을 넣는건가요?

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java [hunk 2]
```diff
@@ -27,6 +28,13 @@ private TokenPayload(Long userId, String deviceId, Date issuedAt) {
         this.userId = userId;
         this.deviceId = deviceId;
         this.issuedAt = issuedAt;
+        this.expiredAt = expiredAt;
+    }
+
+    public void setExpiredAt(Date expiredAt) {
+        if (this.expiredAt == null) {
+            this.expiredAt = expiredAt;
+        }
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java:37]: 원하시는 로직이 'null인 경우에만 입력값으로 설정한다'이면 null이 아닌 경우는 exception을 던지거나 하는 식으로 처리해줘야하지않을까요?

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java [hunk 3]
```diff
@@ -35,27 +36,33 @@ public Token signIn(Email email, String password, String deviceId) {
...
             throw new InvalidTokenException("Access token and refresh token are not matched");
         }
-        if (!deviceCredentialService.isLogin(payload.getUserId(), deviceId)) {
-            throw new LogoutDeviceException();
-        }
 
-        Token newToken = issue(payload.getUserId(), deviceId);
-        deviceCredentialService.updateRefreshToken(payload.getUserId(), deviceId, newToken.getRefreshToken());
+        Token newToken = issue(payload.getUserId(), payload.getDeviceId());
+        TokenPayload newPayload = jwtProvider.getPayload(newToken.getRefreshToken());
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java:60]: token과 payload는 왜 따로 있는건가요?

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java [hunk 4]
```diff
@@ -71,6 +78,7 @@ private Token issue(Long userId, String deviceId) {
             .issuedAt(new Date())
             .build();
 
+        System.out.println("AuthService.issue() payload.getExpiredAt(): " + payload.getExpiredAt());
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java:81]: `System.out.println`은 지양해주세요

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/DeviceCredentialService.java [hunk 5]
```diff
@@ -13,20 +17,47 @@ public class DeviceCredentialService {
...
 
-    public void login(Long userId, String deviceId, String refreshToken) {
+    public void login(Long userId, String deviceId, String refreshToken, LocalDateTime expiredAt) {
+        DeviceCredential deviceCredential = get(userId, deviceId);
+
+        deviceCredential.login(refreshToken, expiredAt);
+        deviceCredentialOutputPort.update(deviceCredential);
     }
 
     public void logout(Long userId, String deviceId) {
```
Comment (f-lab-jason) [inline snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/DeviceCredentialService.java:45]: logout과 delete는 어떻게 다른가요?

**Review by f-lab-jason — COMMENTED** [review]
음 전체적인 로그인 관련해서 세션때 한 번 설명해주시면 좋을 것 같아요. 불필요한 파트들이 너무 많은데다 output port들의 용도도 명확해 보이지 않아서 좀 헷갈려요 😢

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/DeviceCredentialService.java (hunk 5 above)
Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/DeviceCredentialService.java:45]: logout
로그아웃한 시간 및 로그인 상태를 업데이트하고, 저장된 refresh token을 삭제합니다.

delete
디바이스가 삭제될 때 연관된 device credential 데이터도 함께 삭제합니다. 그 때 사용됩니다.

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java (hunk 4 above)
Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java]: 헉 잠깐 확인하려고 넣은 코드인데 삭제를 안했습니다 바로 삭제하겠습니다

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java (hunk 3 above)
Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/service/AuthService.java]: token에서 사용한 정보들을 저장하는 객체를 따로 생성했습니다. 
token을 생성할때나 token에서 정보를 파싱할 때, 어떤 데이터를 사용하는지 명시적으로 드러내는 용도라고 보시면 될 것 같아요.

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java (hunk 2 above)
Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/TokenPayload.java]: [결론] 맞습니다: exception까지 날릴 필요는 없지 않을까? -> 의도와 다른 행동을 하면 exception이 필요하지...

토큰 생성 시 다른 값들은 모두 토큰을 필요로하는 곳에서 정보를 넘겨주는데, expiredAt의 경우에는 환경변수가 필요하기 때문에 jwtProvider에서 넣어주게 됩니다.
원래 setter는 없어야하는 필드이지만 초기화가 필요하여 생성하였고, 초기화를 하는 경우가 없는 경우면 그냥 없는 메서드 취급하면 되지 않을까? 불렀을 때 아무 동작도 안하는 걸로 충분하지. 라는 생각을 했던 것 같습니다.
하지만 저 혼자 만드는 게 아니라 협업이라 생각하면, 해당 메서드가 존재하는 한 다른 이가 값을 바꾸고자하는 의도를 갖고 해당 메서드를 사용할 수 있고, 예외가 발생하지 않는데 원하는 동작이 수행되지 않으면 원인을 찾기 힘들 것 같네요. 
의도가 존재한다면, 해당 의도가 아닌 경우에 대한 처리도 필요한 게 당연한데 생각이 짧았습니다.

File: snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/JwtProvider.java (hunk 1 above)
Comment (sueelly) [reply snapnow-application/src/main/java/kr/flab/snapnow/application/auth/jwt/JwtProvider.java:43]: 존재하지 않는 임의의 값이라는 게 무슨말인지 정확히 모르겠습니다.
payload에는 expiredAt 필드가 존재하고, 생성할 때 해당 값을 설정합니다.
